### PR TITLE
Update sharded_moe.py to support top2 gate with Tutel

### PR DESCRIPTION
**Pull request by xenshinu**
Given the fact that multiple experts per token is very common, and the gather and scatter operation without Tutel is so inefficient, I added support of tutel to top2 gate and tested on pipeline engine. This can be done for any k actually, I'll push that later when I have time to test,

### CODE REVIEW

**Comment by xenshinu** [comment]
@microsoft-github-policy-service agree company="University of Michigan"

**Comment by xenshinu** [comment]
> @xenshinu please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="University of Michigan"

**Comment by xenshinu** [comment]
Not sure if this check is needed
```
    if use_tutel:
        # Tutel doesn't support index values masked with zero
        # so we need to replace masked indices with -1
        indices1_mask = mask1.sum(dim=1) * num_experts - 1
        indices1_s = torch.min(indices1_s, indices1_mask)
        indices2_mask = mask2.sum(dim=1) * num_experts - 1
        indices2_s = torch.min(indices2_s, indices2_mask)
```
I see that in `top1gate`, but when I refer to examples from Tutel https://github.com/microsoft/Tutel/blob/ab7937bb929bc78111d74261b490da25657a7e5c/tutel/impls/fast_dispatch.py#L143, I didn't see any specify check for non-zero mask.

File: deepspeed/moe/sharded_moe.py [hunk 1]
```diff
@@ -517,7 +535,7 @@ def forward(self,
 
         elif self.k == 2:
             gate_output = top2gating(logits, self.capacity_factor if self.training else self.eval_capacity_factor,
-                                     self.min_capacity, self.drop_tokens, self.ep_group, self.top2_2nd_expert_sampling)
+                                     self.min_capacity, self.drop_tokens, self.ep_group, self.top2_2nd_expert_sampling, use_tutel)
```
Comment (loadams) [inline deepspeed/moe/sharded_moe.py:538]: @xenshinu - thanks for this PR, could you run the pre-commit formatter on your branch to resolve the "Formatting" error?  I believe it just wants the `use_tutel` on a new line here.

Comment (xenshinu) [reply deepspeed/moe/sharded_moe.py]: Thanks for pointing that out. I've updated the file and looks like the pre-commit check passed.
Can someone take a look to this question?
https://github.com/microsoft/DeepSpeed/pull/6948#issuecomment-2591036499

Comment (loadams) [reply deepspeed/moe/sharded_moe.py]: @xenshinu - it looks like there are still other errors the formatter is catching, could you try again?  Thanks!

Comment (xenshinu) [reply deepspeed/moe/sharded_moe.py]: I see. I will finish topK soon and deal with the formatting issue together.

**Comment by loadams** [comment]
Hi @xenshinu - FYI you'll need to update from the CLA to DCO requirements in this PR if you're able to?

**Comment by xenshinu** [comment]
> Hi @xenshinu - FYI you'll need to update from the CLA to DCO requirements in this PR if you're able to?

So I need to add signed-and-off in my commit message right? I will do it later. BTW, I ran the pre-commit formatter but it still didn't pass the check, is there any step I missed?

**Comment by loadams** [comment]
> > Hi @xenshinu - FYI you'll need to update from the CLA to DCO requirements in this PR if you're able to?
> 
> Just reset the code to the previous commit and add my signed-and-off.
> 
> @loadams Could you please take care of this PR?

@xenshinu - looks like the formatting check is failing, could you run the pre-commit formatter?  `pre-commit run --all-files` and push those changes?

**Comment by loadams** [comment]
> > > Hi @xenshinu - FYI you'll need to update from the CLA to DCO requirements in this PR if you're able to?
> > 
> > 
> > Just reset the code to the previous commit and add my signed-and-off.
> > @loadams Could you please take care of this PR?
> 
> @xenshinu - looks like the formatting check is failing, could you run the pre-commit formatter? `pre-commit run --all-files` and push those changes?

@xenshinu - curious if you would be able to follow up on this?

**Comment by xenshinu** [comment]
> > > > Hi @xenshinu - FYI you'll need to update from the CLA to DCO requirements in this PR if you're able to?
> > > 
> > > 
> > > Just reset the code to the previous commit and add my signed-and-off.
> > > @loadams Could you please take care of this PR?
> > 
> > 
> > @xenshinu - looks like the formatting check is failing, could you run the pre-commit formatter? `pre-commit run --all-files` and push those changes?
> 
> @xenshinu - curious if you would be able to follow up on this?

Thanks for your reminder. I'll push the TopK version soon and finish the format issue.
